File: easy/array/rotate_array/rotate_array.py
class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        def reverse(arr, start, end):
            while start < end:
                arr[start], arr[end] = arr[end], arr[start]
                start, end = start+1, end-1

        elements = len(nums)
        k %= elements
        reverse(nums, 0, elements-1)
        reverse(nums, 0, k-1)
        reverse(nums, k, elements-1)
        return


s = Solution()

nums = [1, 2, 3, 4, 5, 6, 7]
s.rotate(nums, 3)
print("Solution 1 : ", nums)

nums = [-1, -100, 3, 99]
s.rotate(nums, 2)
print("Solution 2 : ", nums)

nums = [1, 2]
s.rotate(nums, 3)
print("Solution 3 : ", nums)

nums = [1, 2, 3]
s.rotate(nums, 2)
print("Solution 4 : ", nums)

File: easy/array/rotate_array/rotate_array.go
package main

import (
	"fmt"
)

func reverse(arr []int) {
	for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
		arr[i], arr[j] = arr[j], arr[i]
	}
}

func rotate(nums []int, k int) {
	if k == 0 || len(nums) == 1 {
		return
	}
	reverse(nums)

	k %= len(nums)
	firstPart := nums[:k]
	lastPart := nums[k:]

	reverse(firstPart)
	reverse(lastPart)
}

func main() {
	arr := []int{1, 2, 3, 4, 5, 6, 7}
	rotate(arr, 3)
	fmt.Println("Solution 1 : ", arr)

	arr = []int{-1, -100, 3, 99}
	rotate(arr, 2)
	fmt.Println("Solution 2 : ", arr)

	arr = []int{1, 2}
	rotate(arr, 3)
	fmt.Println("Solution 3 : ", arr)

	arr = []int{1, 2, 3}
	rotate(arr, 2)
	fmt.Println("Solution 4 : ", arr)
}
